### PR TITLE
Separate parsing and execution of vernacular

### LINKED ISSUE #15409: Separation of parsing and execution
As mentioned in previous discussions, I made some technical experiments which make me believe it is possible to separate parsing and execution phases in Coq, at reasonable cost. This is a meta-issue explaining the approach and tracking progress.

# Why would we want that?

IDEs often need to reason on ASTs. Since execution of documents is often costly in Coq, having to wait in interaction loops degrades the experience. The typical behavior is: you work on a large file, in which you can navigate without trouble. Then you add a `Notation` or an `Import` command in the middle of the file. Suddenly, editing anything above the command requires re-executing before you can work again on the bottom of the document (or at least get useful feedback from the IDE, in a non-blocking model).

From the implementation standpoint, we anyway had to classify commands with parsing effects. But we don't have the proper separation of state that makes it possible to execute these effects without a full execution. In a way, we already pay the complexity but we don't get the benefits.

Other tools than IDEs can benefit greatly from being able to reliably parse Coq documents without executing them. For example, `coqdoc` could just call the parsing API.

# Why is it not so easy?

Remember that Coq has extensible parsing, meaning that some commands can change the grammar. So by a parsing phase, we really mean a _syntactic interpretation_ phase, where we parse text into an AST _and_ maintain the grammar up to date at each point of the document.

# What properties do we want?

Essentially, we want to refine the current vernac interpretation API:

```ocaml
val interp : Vernacstate.t -> Vernacexpr.vernac_control -> Vernacstate.t
```

into:

```ocaml
val synterp : Synterp.State.t -> Vernacexpr.vernac_control -> Synterp.State.t
val interp : Synterp.State.t -> Vernacstate.t -> Vernacexpr.vernac_control -> Vernacstate.t
```

Note that `interp` reads the syntactic interpretation state, but cannot write to it.

The API can be used for example in three ways:
- Folding `synterp` on a document, to get a full AST without executing the document (for example, for `coqdoc`).
- For each sentence, running `synterp` followed by `interp` (for example, for `coqtop`).
- Folding `synterp` on a document, then executing some commands by running `interp` (for example, for IDEs).

# What kind of patches are required?

- The `Notation` command interpretation has to be split in two (one for grammar extension, the other for notation interpretation) https://github.com/coq/coq/pull/15755
- `where` notations should not declare grammar extensions https://github.com/coq/coq/pull/15468
- Nametab state has to be split in two (syntactic interpretation needs to be able to resolve module names) https://github.com/coq/coq/pull/15756
- The summary needs a way to distinguish the part of the system state corresponding to the syntactic interpretation phase
- Module internalization needs to be separated from their interpretation, which requires term typing https://github.com/coq/coq/pull/15500
- Libstack and libobject traversal has to be separated in two, one for each phase
- Parsing effects of `Declare ML modules` on plugins declaring grammar rules have to be separated
- Vernac extensions need to be able to declare commands whose execution has parsing effects (example: `Tactic Notation`)
- `synterp` needs to be implemented as a new dispatch on vernacular expressions

# What could we try next?

Being able to resolve names without fully executing a document would also be a nice feature for IDEs and tools. We could study that next.

### PR DESCRIPTION
**Pull request by maximedenes**
Closes #15409.

TODO:

- [x] Overlay for elpi_hb
- [x] Overlay for lean_importer
- [x] Overlay for serapi
- [x] Investigate slowdown
- [x] Add some dynamic checks
- [x] Open PRs for overlays

Overlays:
- https://github.com/coq-tactician/coq-tactician/pull/62
- https://github.com/ejgallego/coq-serapi/pull/327
- https://github.com/MetaCoq/metacoq/pull/895
- https://github.com/SkySkimmer/coq-lean-import/pull/7
- https://github.com/ejgallego/coq-lsp/pull/478
- https://github.com/LPCIC/coq-elpi/pull/447

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@coqbot bench

File: vernac/vernacentries.ml [hunk 1]
```diff
@@ -2440,55 +2258,34 @@ let translate_vernac ?loc ~atts v = let open Vernacextend in match v with
...
     vtdefault(fun () -> vernac_coercion ~atts r st)
+
   | VernacIdentityCoercion ({v=id},s,t) ->
     vtdefault(fun () -> vernac_identity_coercion ~atts id s t)
 
   (* Type classes *)
   | VernacInstance (name, bl, t, props, info) ->
     let atts, program = Attributes.(parse_with_extra program) atts in
-    if program then
+    (if program then
```
Comment (SkySkimmer) [inline vernac/vernacentries.ml:2288]: ?

File: vernac/vernacentries.ml [hunk 2]
```diff
@@ -2650,11 +2449,9 @@ let translate_vernac ?loc ~atts v = let open Vernacextend in match v with
     let using = Option.append using (Proof_using.get_default_proof_using ()) in
     let tacs = if Option.is_empty tac then "tac:no" else "tac:yes" in
     let usings = if Option.is_empty using then "using:no" else "using:yes" in
-    Aux_file.record_in_aux_at "VernacProof" (tacs^" "^usings);
+    Aux_file.record_in_aux_at "EVernacProof" (tacs^" "^usings);
```
Comment (SkySkimmer) [inline vernac/vernacentries.ml:2452]: It's just VernacProof though?

File: vernac/vernacinterp.ml [hunk 3]
```diff
@@ -184,34 +201,27 @@ and vernac_load ~verbosely fname =
...
-      let stack, pm = v_mod (interp_control ~st:{ st with Vernacstate.lemmas = stack; program = pm }) stm in
+    | Some { loc; v = cmd } ->
+      let interp_fn ?loc ~st ~atts expr =
+        let e = Synterp.synterp ?loc ~atts expr in
+        interp_expr ?loc ~atts ~st e
+      in
+      (*
+      let expr = Synterp.synterp ?loc ~atts:stm.attrs stm.expr in
+      let stm = CAst.make ?loc { attrs = stm.attrs; control = stm.control; expr } in
+      *)
```
Comment (SkySkimmer) [inline vernac/vernacinterp.ml:212]: What does this comment mean?

File: vernac/vernacinterp.ml [hunk 4]
```diff
@@ -184,34 +201,27 @@ and vernac_load ~verbosely fname =
...
+      let stack, pm = v_mod (interp_control ?loc ~st:{ st with Vernacstate.interp = { st.interp with Vernacstate.Interp.lemmas = stack; program = pm }} ~interp_fn ~atts:cmd.attrs) (cmd.control,cmd.expr) in
```
Comment (SkySkimmer) [inline vernac/vernacinterp.ml:213]: This line is getting a bit long

File: vernac/vernacexpr.ml [hunk 5]
```diff
@@ -477,10 +482,15 @@ type nonrec vernac_expr =
...
-  | VernacProofMode of string
 
-  (* For extension *)
-  | VernacExtend of extend_name * Genarg.raw_generic_argument list
+  | VernacAddOption of Goptions.option_name * Goptions.table_value list
+  | VernacRemoveOption of Goptions.option_name * Goptions.table_value list
+
+type 'a vernac_expr_gen =
+  | VernacSynterp of 'a
+  | VernacPure of pure_vernac_expr
```
Comment (SkySkimmer) [inline vernac/vernacexpr.ml:491]: I don't think this name is correct, it is not generally pure but only pure wrt syntax. Maybe 
```suggestion
  | VernacSynPure of synpure_vernac_expr
```
?

File: vernac/vernacexpr.ml [hunk 6]
```diff
@@ -343,17 +343,51 @@ type hints_expr =
...
+
+  | VernacSetOption of bool (* Export modifier? *) * Goptions.option_name * option_setting
+  | VernacProofMode of string
+
+  (* For extension *)
+  | VernacExtend of extend_name * Genarg.raw_generic_argument list
+
+type nonrec pure_vernac_expr =
 
   | VernacLoad of verbose_flag * string
```
Comment (SkySkimmer) [inline vernac/vernacexpr.ml:385]: Load is not synpure.
I guess we should have 
~~~ocaml
type synterp_vernac_expr =
  | VernacLoad of verbose_flag * string
...
~~~
and something like
~~~ocaml
type synterp_entry =
  | VernacLoad of verbose_flag * vernac_entry_control list
...
~~~

**Review by SkySkimmer — DISMISSED** [review]
VernacPure is not the right name, and Load is bugged.

Also not sure what's going on with coqdoc in the test suite.

File: library/global.ml [hunk 7]
```diff
@@ -57,12 +57,12 @@ let globalize f =
 
 let globalize0_with_summary fs f =
   let env = f (safe_env ()) in
-  Summary.unfreeze_summaries ~partial:true fs;
+  Summary.unfreeze_summaries fs;
```
[the remark this answers is not in the document]
Comment (maximedenes) [reply library/global.ml:60]: The `partial:true` was introduced in a previous PR separating the state, but wasn't fully correct (we don't do the same thing on non-present summaries depending on the value of `partial`). This PR fixes it by making the type of summaries for a given stage more precise and removes the reliance on `partial:true`.

File: vernac/vernacinterp.ml (hunk 3 above)
Comment (maximedenes) [reply vernac/vernacinterp.ml:212]: It means you're looking at a draft PR ;)

File: library/global.ml (hunk 7 above)
[the remark this answers is not in the document]
Comment (SkySkimmer) [reply library/global.ml:60]: There's still `partial:true` in stm though

**Comment by maximedenes** [comment]
> Also not sure what's going on with coqdoc in the test suite.

Probably the `.glob` emitting for sections has to be done at synterp time if we want to preserve the current behavior.

File: vernac/vernacexpr.ml (hunk 5 above)
Comment (SkySkimmer) [reply vernac/vernacexpr.ml:491]: It's not about anticipating, it's about correctness today

Comment (maximedenes) [reply vernac/vernacexpr.ml:491]: Today we distinguish only syntax effects at that place. So pure = no syntax effect.

File: vernac/metasyntax.ml [hunk 8]
```diff
@@ -917,6 +917,7 @@ let classify_syntax_definition (local, _) =
   if local then Dispose else Substitute
 
 let open_syntax_extension i o =
+  (* Printf.eprintf "open_syntax_extension called\n"; *)
```
[the remark this answers is not in the document]
Comment (maximedenes) [reply vernac/metasyntax.ml]: Fixed.

File: vernac/vernacentries.ml (hunk 1 above)
Comment (maximedenes) [reply vernac/vernacentries.ml]: Fixed.

File: vernac/vernacentries.ml (hunk 2 above)
Comment (maximedenes) [reply vernac/vernacentries.ml]: Fixed.

File: vernac/vernacinterp.ml (hunk 3 above)
Comment (maximedenes) [reply vernac/vernacinterp.ml]: Fixed.

File: vernac/vernacinterp.ml (hunk 4 above)
Comment (maximedenes) [reply vernac/vernacinterp.ml]: Fixed.

File: vernac/vernacexpr.ml (hunk 5 above)
Comment (maximedenes) [reply vernac/vernacexpr.ml]: Name changed.

File: vernac/vernacexpr.ml (hunk 6 above)
Comment (maximedenes) [reply vernac/vernacexpr.ml:384]: Yes, but I need to think a bit more about control there.

**Comment by Alizter** [comment]
That is some significant overhead. What is going on?

File: vernac/vernacentries.mli [hunk 9]
```diff
@@ -19,12 +19,13 @@ val check_may_eval :
...
   -> atts:Attributes.vernac_flags
-  -> Vernacexpr.vernac_expr
+  -> Synterp.vernac_entry
   -> Vernacextend.typed_vernac
 
 (** Vernacular require command, used by the command line *)
 val vernac_require
-  : Libnames.qualid option
+  : Library.library_t list
+  -> Names.DirPath.t list
```
Comment (SkySkimmer) [inline vernac/vernacentries.mli:28]: Maybe we should keep this as the combined version and add `vernac_interp_require` if useful?

File: vernac/synterp.mli [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+type synterp_entry =
+  | EVernacNoop
+  | EVernacNotation of bool * Metasyntax.notation_interpretation_decl
+  | EVernacBeginSection of Names.lident
+  | EVernacEndSegment of Names.lident
+  | EVernacRequire of
+      Library.library_t list * DirPath.t list * Vernacexpr.export_with_cats option * (qualid * Vernacexpr.import_filter_expr) list
+  | EVernacImport of (Vernacexpr.export_flag *
+      Libobject.open_filter) *
```
Comment (SkySkimmer) [inline vernac/synterp.mli:53]:
```suggestion
  | EVernacImport of (Vernacexpr.export_flag * Libobject.open_filter) *
```

File: vernac/synterp.mli [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+  atts:Attributes.vernac_flags -> (module_local:bool -> 'a) -> 'a
+
+val with_def_attributes :
+  ?coercion:bool -> atts:Attributes.vernac_flags -> (atts:DefAttributes.t -> 'a) -> 'a
+
+type module_entry = Modintern.module_struct_expr * Names.ModPath.t * Modintern.module_kind * Entries.inline
+
+type synterp_entry =
+  | EVernacNoop
+  | EVernacNotation of bool * Metasyntax.notation_interpretation_decl
```
Comment (SkySkimmer) [inline vernac/synterp.mli:47]: inline record?

File: vernac/vernacentries.mli (hunk 9 above)
Comment (SkySkimmer) [reply vernac/vernacentries.mli:28]: I guess the split version would be useful for `coqc -only-parsing -require foo`

File: tactics/autorewrite.ml [hunk 12]
```diff
@@ -537,7 +537,7 @@ let find_applied_relation ?loc env sigma c left2right =
                        spc () ++ str"of this term does not end with an applied relation.")
 
 let default_hint_rewrite_locality () =
-  if Global.sections_are_opened () then Hints.Local
+  if Lib.sections_are_opened () then Hints.Local
```
[the remark this answers is not in the document]
Comment (maximedenes) [reply tactics/autorewrite.ml:543]: `Lib.sections_are_opened` does not depend on typing.

File: vernac/vernacexpr.ml (hunk 6 above)
Comment (maximedenes) [reply vernac/vernacexpr.ml]: Thanks for the suggestion, this is now done.

File: vernac/synterp.mli (hunk 10 above)
Comment (maximedenes) [reply vernac/synterp.mli]: Done, thanks!

File: vernac/synterp.mli (hunk 11 above)
Comment (maximedenes) [reply vernac/synterp.mli]: Done, thanks!

**Comment by SkySkimmer** [comment]
>INFO: failed to install coq-compcert
coq-corn

logs say the "fetch" failed, I guess the network had a temporary issue

File: vernac/vernacentries.mli (hunk 9 above)
Comment (maximedenes) [reply vernac/vernacentries.mli:35]: Done.

**Comment by maximedenes** [comment]
@coqbot bench

**Review by jfehrle — COMMENTED** [review]
This is good, thanks for doing this.

A couple thoughts.
- What interface would be exposed to IDEs to pre-parse a file?  Would IDEs be able to send an entire file to be parsed at once without executing it or would they have to send it incrementally?
- Have you measured how long it takes to parse large files, e.g. `Pff.v`?  Just to give a rough idea of the performance (and how often an IDE could reparse).
- Is there a way (or a plan or a hope) to recover from parsing errors to give a better IDE user experience?  It would also be nice to be able to report errors such as references to nonexistent theorems, ltac tactics, etc.  Would that fit into the scheme of this PR?
- How about providing a few code comments explaining `VernacSynterp` and maybe a reference to #15409?

**Comment by maximedenes** [comment]
> * What interface would be exposed to IDEs to pre-parse a file?

What do you mean by "pre-parse"?

> * Would IDEs be able to send an entire file to be parsed at once without executing it or would they have to send it incrementally?

The API is geared towards incremental parsing (i.e. sentence by sentence). But in practice, it is very easy to iterate over a file, for example the first time it is opened.

> * Have you measured how long it takes to parse large files, e.g. `Pff.v`?

I don't have precise measurements at hand, but we tested in VsCoq 2 (which uses this separation patch), and it takes a few seconds IIRC. `Pff` doesn't have many commands with parsing effects, so this patch should not add much to what was already necessary for pure parsing of the file.

> * Is there a way (or a plan or a hope) to recover from parsing errors to give a better IDE user experience?

Today recovering at the sentence level is fairly easy to do (it's essentially a one-liner looking for the next sentence terminator). Having error resiliency (as opposed to error recovery) is more difficult, but would probably be necessary for e.g. completion on partial sentences, which is an interesting use case. Not directly related to this patch.

> * It would also be nice to be able to report errors such as references to nonexistent theorems, ltac tactics, etc.  Would that fit into the scheme of this PR?

I would love to have that, but name resolution is dependent on full typing in Coq as it is today. So unlike parsing, having a separate name resolution phase would really mean re-architecturing the system. Here we only propagate an existing invariant (parsing does not depend on typing) to the API.

> * How about providing a few code comments explaining `VernacSynterp` and maybe a reference to #15409?

Yes, I will add some explanations in the code.

**Comment by gares** [comment]
> I would love to have that, but name resolution is dependent on full typing in Coq as it is today. So unlike parsing, having a separate name resolution phase would really mean re-architecturing the system. Here we only propagate an existing invariant (parsing does not depend on typing) to the API.

I promised myself to give this a shot at some point, but first we need to finish separating this phase.

**Comment by jfehrle** [comment]
>> What interface would be exposed to IDEs to pre-parse a file?

> What do you mean by "pre-parse"?

For example, to parse an entire file to populate a list of theorems defined in it before the user executes any of it.

**Comment by gares** [comment]
It depends what you want for the statements.
If you are OK with statements that parse, but not necessarily type check, then this PR (stream of PRS) is enough.

File: vernac/vernacstate.ml [hunk 13]
```diff
@@ -118,20 +168,34 @@ let do_if_not_cached rf f v =
...
     program = !s_program;
     opaques = Opaques.Summary.freeze ~marshallable;
     shallow = false;
-    parsing = Parser.cur_state ();
   }
 
-let unfreeze_interp_state { system; lemmas; program; parsing; opaques } =
-  do_if_not_cached s_cache System.unfreeze system;
+let unfreeze_interp_state { system; lemmas; program; opaques } =
+  do_if_not_cached s_cache System.Interp.unfreeze system;
```
[the remark this answers is not in the document]
Comment (maximedenes) [reply vernac/vernacstate.ml:179]: I tested by implementing a similar cache for synterp, I didn't bring anything so I removed it.

**Comment by SkySkimmer** [comment]
@coqbot bench

File: library/goptions.mli [hunk 14]
```diff
@@ -192,6 +192,9 @@ type table_value =
   | StringRefValue of string
   | QualidRefValue of Libnames.qualid
 
+(** [get_option_value key] returns [None] if option with name [key] was not found. *)
+val get_option_value : option_name -> option_value option
```
Comment (SkySkimmer) [inline library/goptions.mli:196]: Maybe this should be 
```suggestion
val get_option_value : option_name -> (unit -> option_value) option
```
instead? Then the metacoq code would look like
~~~ocaml
let timing_opt =
  let open Goptions in
  let key = ... in
  match get_option_value key with
  | Some get -> get
  | None -> declare_bool_and_ref ... (* no () argument *)
~~~

**Review by SkySkimmer — APPROVED** [review]
This looks about ready modulo the minor comment on goptions IMO

File: library/goptions.mli (hunk 14 above)
Comment (maximedenes) [reply library/goptions.mli:196]: I don't have a strong opinion, and can easily change this. But doesn't the current API look more natural than the one you suggest, if you forget about the specific Metacoq use case?

Comment (SkySkimmer) [reply library/goptions.mli:196]: I feel the metacoq usage is a bit of a phase separation violation, even if it's a benign one. The `let _ = timing_opt () (* just for init *)` is both ugly and fetches a Interp-stage value at plugin link time.

**Comment by maximedenes** [comment]
@coqbot run full CI

**Comment by maximedenes** [comment]
@coqbot run full CI

File: default.nix [hunk 15]
```diff
@@ -81,6 +81,17 @@ stdenv.mkDerivation rec {
...
+      ppx_sexp_conv
+      ppx_deriving_yojson
+      ppx_hash
+      ppx_compare
+      sexplib
+      cmdliner
+      dune-build-info
+      menhir
+      uri
+      elpi
```
Comment (Zimmi48) [inline default.nix:94]: What's up with all of this? Is this an accidental introduction linked to you needing these dependencies to prepare the overlays?

Comment (maximedenes) [reply default.nix:94]: Argh! This is a spurious commit, will fix, sorry for the noise.

**Comment by SkySkimmer** [comment]
I'll merge tomorrow assuming no further comments or problems

File: default.nix (hunk 15 above)
Comment (maximedenes) [reply default.nix]: Removed.

**Pull request merged**